server/service/src/document/document_service.rs

```diff
@@ -0,0 +1,440 @@
+                }
+            })
+            .map_err(|err| err.to_inner_error())?;
+        Ok(document)
+    }
+}
+
+pub struct DocumentService {}
+impl DocumentServiceTrait for DocumentService {}
+
+impl From<RepositoryError> for DocumentInsertError {
+    fn from(err: RepositoryError) -> Self {
+        match err {
+            RepositoryError::NotFound => DocumentInsertError::InvalidDocumentHistory,
+            _ => DocumentInsertError::DatabaseError(err),
+        }
+    }
+}
+
+fn two_way_document_merge(our: RawDocument, their: Document) -> RawDocument {
+    let our_data = our.data;
+    let their_data = their.data;
+
+    let solver = TakeLatestConflictSolver::new(our.timestamp.clone(), their.timestamp.clone());
+    let merged = two_way_merge(&our_data, &their_data, &solver);
+
+    let mut new_parents = our.parents;
+    new_parents.push(their.id);
+
+    RawDocument {
+        parents: new_parents,
+        timestamp: Utc::now(),
+        data: merged,
+
+        // keep exiting
+        name: our.name,
+        r#type: our.r#type,
+        author: our.author,
+        schema_id: our.schema_id,
+    }
+}
+
+fn three_way_document_merge(our: RawDocument, their: Document, base: Document) -> RawDocument {
+    let our_data = our.data;
+    let their_data = their.data;
+    let base_data = base.data;
+
+    let solver = TakeLatestConflictSolver::new(our.timestamp.clone(), their.timestamp.clone());
+    let merged = three_way_merge(&our_data, &their_data, &base_data, &solver);
+
+    let mut new_parents = our.parents;
+    new_parents.push(their.id);
+
+    RawDocument {
+        parents: new_parents,
+        timestamp: Utc::now(),
+        data: merged,
+
+        name: our.name,
+        author: our.author,
+        r#type: our.r#type,
+        schema_id: our.schema_id,
+    }
+}
+
+fn json_validator(
+    connection: &StorageConnection,
+    doc: &RawDocument,
+) -> Result<Option<JSONSchema>, DocumentInsertError> {
+    if let Some(schema_id) = &doc.schema_id {
+        let schema_repo = JsonSchemaRepository::new(connection);
+        let schema = schema_repo.find_one_by_id(schema_id)?;
+        let compiled = match JSONSchema::compile(&schema.schema) {
+            Ok(v) => Ok(v),
+            Err(err) => Err(DocumentInsertError::InternalError(format!(
+                "Invalid json schema: {}",
+                err
+            ))),
+        }?;
+        return Ok(Some(compiled));
+    }
+    Ok(None)
+}
+
+fn validate_json(validator: &JSONSchema, data: &serde_json::Value) -> Result<(), Vec<String>> {
+    Ok(validator.validate(data).map_err(|errors| {
+        let errors: Vec<String> = errors.into_iter().map(|err| format!("{}", err)).collect();
+        errors
+    })?)
+}
+
+/// Does a raw insert without schema validation
+fn insert_document(
+    connection: &StorageConnection,
+    store_id: &str,
+    doc: RawDocument,
+) -> Result<Document, DocumentInsertError> {
+    let repo = DocumentRepository::new(connection);
```

Comment:
This would be a lot easier if documents were storing field changes:

```rust
/* use correct abstraction for the following */

let documents = /* get all documents for name */
let ordered_documents = /* order by timestamp */

let current_document_state = 
/* iterate over ordered_documents */
     /* for each field change that is in current schema */
        /* current_document_state[field_name] = field_value */

let all_parent_references = 
/* map over documents -> parent_id */
let parents = 
/* filter documents -> id not in all_parent_references */

let changes = 
/* filter fields in data from input -> doesn't match current_document_state */

let new document = /* parents, changes, ... */ 
```

this would be merge/consolidation logic in it's entirety, with `current_document_state` getter being used when getting document

Reply:
this is slightly different, parent_ids do actually need to be specified. 

revised

```rust
/* use correct abstraction for the following */

let documents = /* get all documents for document name */
let documents = /* filter documents for just parent_ids and ancestors of those parent_ids */

let ordered_documents = /* order by timestamp */

let current_document_state = 
/* iterate over ordered_documents */
     /* for each field change that is in current schema */
        /* current_document_state[field_name] = field_value */

let changes = 
/* filter fields in data from input -> doesn't match current_document_state */

let new document = /* changes, ... */ 
```

Reply:
you also need to consider field deletion (not a big thing to add, just mentioning)

If we use an additional logical clock (like a Lamport timestamp) we can also fix the merge edge case. (We should use an additional logical clock in any case to have fast and correct topological sorts.)

Some pros /cons:

What about performance? since you have to apply ALL diffs every time.

Can you apply different merge strategies? e.g. for merging arrays? or are we restricted to a "TakeLatestSolver"?

Just had a look at the existing code again and its basically:
``` rust
let base = common_ancestor(ours, theirs);
let new_doc = merge(base, ours, theirs);
``` 
With merge and common_ancestor being reasonably simple/small. With all the details added for the proposed diff version its actually not too much difference.

With the "standard" way of handling merges you know that its scalable. You are also more flexible how you store your data, e.g. a hybrid of snapshots and diffs. An initial version wouldn't need to handle diffs at all and it can be added as a later optimization.

Reply:
> you also need to consider field deletion (not a big thing to add, just mentioning)

I think it's there `/* for each field change that is in current schema */`

> What about performance? since you have to apply ALL diffs every time.

I'll bench it

> Can you apply different merge strategies? e.g. for merging arrays? or are we restricted to a "TakeLatestSolver"?

Yeah `/* current_document_state[field_name] = field_value */` can be modified by type

> Just had a look at the existing code again and its basically:

Isn't that over simplifying things ? The `merge` 

> With merge and common_ancestor being reasonably simple/small

I don't think this statement is accurate for `merge` (if you follow down the chain of helper methods)

> With the "standard" way of handling merges you know that its scalable. You are also more flexible how you store your data, e.g. a hybrid of snapshots and diffs. An initial version wouldn't need to handle diffs at all and it can be added as a later optimization.

I am really worried that this will be write only code. 

It looks like I failed convincing you with pseudocode, I'll have to make a real example and we'll have to ask wider team for opinion. (I am still convinced that simpler and more direct way of solving the problem is the way to go)